### ISSUE TITLE
AppDomainWrapper memory leak

### ISSUE THREAD

**Issue by sungam3r** [NLog/NLog#1689]
**Type**: Bug

**NLog version**: 4.3.9

**Platform**: .Net 4.5

- What is the current result? - memory leak of AppDomainWrapper with EventHandler callbacks
- What is the expected result? - lack of (or less) AppDomainWrapper instances in managed heap

This causes memory leak:

1.  subscription 
    appDomain.ProcessExit += OnProcessExit;
    appDomain.DomainUnload += OnDomainUnload;
    in AppDomainWrapper ctor (there is no unsubscription at all)

2. Instance created each time  public static AppDomainWrapper CurrentDomain { get { return new AppDomainWrapper(AppDomain.CurrentDomain); } } method called

3. Example of transient instance creation call in FilePathLayout:
            if (_filePathKind == FilePathKind.Relative)
            {
                _baseDir = AppDomainWrapper.CurrentDomain.BaseDirectory;
            }
![memory](https://cloud.githubusercontent.com/assets/21261007/19188992/34710862-8c9e-11e6-923c-e82be18116a6.png)

**Comment by 304NotModified** [comment]
hey! Any how to fix this? Could you send a PR?

**Comment by snakefoot** [comment]
One solution could be that LogFactory didn't cache the AppDomainWrapper but just returned a new instance every time.

Another solution could be that AppDomainWrapper unsubscribed itself from both handlers when one of the event-handlers fired.

**Comment by snakefoot** [comment]
Yet another solution would be that LogFactory sets currentAppDomain = null when it is disposed. But the dispose-logic in LogFactory looks very tangled. Ex. the reload-timer is disposed every where.

**Comment by 304NotModified** [comment]
@sungam3r please your idea on this / https://github.com/NLog/NLog/pull/1690, thanks!

**Comment by sungam3r** [comment]
Thank you for https://github.com/NLog/NLog/pull/1690
May I ask why not make
```
public interface IAppDomain
{
  event EventHandler ProcessExit;
  event EventHandler DomainUnload;
}
```
and simply delegate event handler subscriptions to underlying AppDomain object in AppDomainWrapper
```
  public event EventHandler ProcessExit
        {
            add
            {
                this.currentAppDomain.ProcessExit += value;
            }
            remove
            {
                this.currentAppDomain.ProcessExit -= value;
            }
        }

        public event EventHandler DomainUnload
        {
            add
            {
                this.currentAppDomain.DomainUnload += value;
            }
            remove
            {
                this.currentAppDomain.DomainUnload -= value;
            }
        }
```
In fact, ProcessExit and DomainUnload events do not need template event args.

**Comment by snakefoot** [comment]
If you look at my commits, then you will see that it was also my initial suggestion, but it would be a breaking change. But this perfect forward should be included in #1691

**Comment by 304NotModified** [comment]
https://github.com/NLog/NLog/pull/1690 has been released in NLog 4.3.10 and so I assume this issue has been fixed. If not, please let us know!
